File: src/MagnusAddon/language_services/janome_ex/word_extraction/word_exclusion.py
from __future__ import annotations

from typing import TYPE_CHECKING

from autoslot import Slots

if TYPE_CHECKING:
    from sysutils.json.json_reader import JsonReader

from typing import Any


class WordExclusion(Slots):
    secret = "aoesunth9cgrcgf"
    _no_index = -1
    def __init__(self, word: str, index: int, _secret:str) -> None:
        if _secret != "aoesunth9cgrcgf": raise ValueError("please use the factory methods instead of this private constructor")
        self.word = word
        self.index = index

    def excludes_form_at_index(self, form: str, index: int) -> bool:
        return form == self.word and (self.index == WordExclusion._no_index or self.index == index)

    @classmethod
    def global_(cls, exclusion: str) -> WordExclusion: return WordExclusion(exclusion.strip(), WordExclusion._no_index, WordExclusion.secret)
    @classmethod
    def at_index(cls, exclusion: str, index: int) -> WordExclusion: return WordExclusion(exclusion.strip(), index, WordExclusion.secret)

    def __eq__(self, other: object) -> bool:
        if isinstance(other, WordExclusion):
            return self.word == other.word and self.index == other.index
        return False

    def __hash__(self) -> int:
        return hash((self.word, self.index))

    def __repr__(self) -> str:
        return f"WordExclusion('{self.word}', {self.index})"

    def excludes_all_words_excluded_by(self, other: WordExclusion) -> bool:
        return self.word == other.word and (self.index == WordExclusion._no_index or self.index == other.index)

    def to_dict(self) -> dict[str, Any]:
        return {"word": self.word, "index": self.index}

    @classmethod
    def from_reader(cls, reader: JsonReader) -> WordExclusion:
        return cls(word=reader.string("word"), index=reader.int("index"), _secret=WordExclusion.secret)

File: src/src_dotnet/JAStudio.Core/LanguageServices/JanomeEx/WordExtraction/WordExclusion.cs
using System;
using System.Collections.Generic;
using JAStudio.Core.SysUtils.Json;

namespace JAStudio.Core.LanguageServices.JanomeEx.WordExtraction;

public sealed class WordExclusion
{
   const int NoIndex = -1;

   public string Word { get; }
   public int Index { get; }

   WordExclusion(string word, int index)
   {
      Word = word;
      Index = index;
   }

   public bool ExcludesFormAtIndex(string form, int index) => form == Word && (Index == NoIndex || Index == index);

   public static WordExclusion Global(string exclusion) => new(exclusion.Trim(), NoIndex);

   public static WordExclusion AtIndex(string exclusion, int index) => new(exclusion.Trim(), index);

   /// <summary>
   /// Creates a WordExclusion from a string in the format "word:index" or "word" for global exclusions.
   /// </summary>
   public static WordExclusion FromString(string value)
   {
      var parts = value.Split(':');
      if(parts.Length == 1)
      {
         return Global(parts[0]);
      } else if(parts.Length == 2)
      {
         return AtIndex(parts[0], int.Parse(parts[1]));
      }

      throw new ArgumentException($"Invalid exclusion format: {value}");
   }

   public override bool Equals(object? obj)
   {
      if(obj is WordExclusion other)
      {
         return Word == other.Word && Index == other.Index;
      }

      return false;
   }

   public override int GetHashCode() => HashCode.Combine(Word, Index);

   public override string ToString() => $"WordExclusion('{Word}', {Index})";

   public bool ExcludesAllWordsExcludedBy(WordExclusion other) => Word == other.Word && (Index == NoIndex || Index == other.Index);

   public Dictionary<string, object> ToDict() =>
      new()
      {
         { "word", Word },
         { "index", Index }
      };

   public static WordExclusion FromReader(JsonReader reader) =>
      new(
         reader.GetString("word"),
         reader.GetInt("index")
      );
}
